sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java

```diff
@@ -0,0 +1,501 @@
+    }
+  }
+
+  /**
+   * A POJO encapsulating a configuration for retry behavior when issuing requests to dynamodb. A
+   * retry will be attempted until the maxAttempts or maxDuration is exceeded, whichever comes
+   * first, for any of the following exceptions:
+   *
+   * <ul>
+   *   <li>{@link IOException}
+   * </ul>
+   */
+  @AutoValue
+  public abstract static class RetryConfiguration implements Serializable {
+    @VisibleForTesting
+    static final RetryPredicate DEFAULT_RETRY_PREDICATE = new DefaultRetryPredicate();
+
+    abstract int getMaxAttempts();
+
+    abstract Duration getMaxDuration();
+
+    abstract DynamoDBIO.RetryConfiguration.RetryPredicate getRetryPredicate();
+
+    abstract DynamoDBIO.RetryConfiguration.Builder builder();
+
+    public static DynamoDBIO.RetryConfiguration create(int maxAttempts, Duration maxDuration) {
+      checkArgument(maxAttempts > 0, "maxAttempts should be greater than 0");
+      checkArgument(
+          maxDuration != null && maxDuration.isLongerThan(Duration.ZERO),
+          "maxDuration should be greater than 0");
+      return new AutoValue_DynamoDBIO_RetryConfiguration.Builder()
+          .setMaxAttempts(maxAttempts)
+          .setMaxDuration(maxDuration)
+          .setRetryPredicate(DEFAULT_RETRY_PREDICATE)
+          .build();
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract DynamoDBIO.RetryConfiguration.Builder setMaxAttempts(int maxAttempts);
+
+      abstract DynamoDBIO.RetryConfiguration.Builder setMaxDuration(Duration maxDuration);
+
+      abstract DynamoDBIO.RetryConfiguration.Builder setRetryPredicate(
+          RetryPredicate retryPredicate);
+
+      abstract DynamoDBIO.RetryConfiguration build();
+    }
+
+    /**
+     * An interface used to control if we retry the BatchWriteItemRequest call when a {@link
+     * Throwable} occurs. If {@link RetryPredicate#test(Object)} returns true, {@link Write} tries
+     * to resend the requests to the dynamodb server if the {@link RetryConfiguration} permits it.
+     */
+    @FunctionalInterface
+    interface RetryPredicate extends Predicate<Throwable>, Serializable {}
+
+    private static class DefaultRetryPredicate implements RetryPredicate {
+      private static final ImmutableSet<Integer> ELIGIBLE_CODES =
+          ImmutableSet.of(HttpStatus.SC_SERVICE_UNAVAILABLE);
+
+      @Override
+      public boolean test(Throwable throwable) {
+        return (throwable instanceof IOException
+            || (throwable instanceof AmazonDynamoDBException)
+            || (throwable instanceof AmazonDynamoDBException
+                && ELIGIBLE_CODES.contains(((AmazonDynamoDBException) throwable).getStatusCode())));
+      }
+    }
+  }
+
+  /** Implementation of {@link #write}. */
+  @AutoValue
+  public abstract static class Write
+      extends PTransform<PCollection<BatchWriteItemRequest>, PCollectionTuple> {
```

Comment:
Why this returns a `PCollectionTuple`? to have multiple outputs?  Is this really needed? Can we achieve the same by returning `PCollection<Void>`?

Reply:
This is to make it consistent with SNSIO. Of course, I can try with Void.

Reply:
Oh wow I think we missed that one in the review of SnsIO. Now I have to check on why it is like that, thanks for the explanation. And thanks for being receptive for the  change suggestions.

Reply:
@iemejia , I have changed it to use `PCollection<Void>`, but couldn't find a way to return multiple outputs. Something similar to this JDBC one, https://github.com/apache/beam/blob/master/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java#L842, it also return just one output if using `PCollection<Void>`. Any idea of how this would be done?

Reply:
Let me look into this and tell you, I don't have a use case at hand but worth checking.

Reply:
Ah wait I have misread this. What you mean with multiple outputs, can't we just use a `PCollectionTuple` and connect each output PCollection to the different outputting IOs?

Reply:
Hi @iemejia , thanks for your response. As we discussed, the `PCollectionTuple` is unnecessary, since we just need to return one output. But implementing with PCollection<Void> will not give us a option to turn any output. I have put together my internal PR for your convenience to review the implementation of PCollection<BatchWriteItemResult> for the Writer, if you don't mind? https://github.com/cmachgodaddy/beam/pull/1/files#diff-127519743356d99d2d217126be4f240eR316

Reply:
Since this is a sink you may not really care about multiple outputs in the return type of the `Write` definition. You can achieve multiple outputs by creating subgraphs in the pipeline with `PCollectionTuple` and then chaining to each outputting path the write transform.

We could have decided to make the signature `PTransform<PCollection<BatchWriteItemRequest>, PDone>` too, but I asked explictly to use the `PCollection<Void>` as a return type to enable in the future the possibility to have patterns like waiting for a PCollection write before another, that's the reason why  (sorry if I was not explicit about this).

I still have to check `SnsIO`  to see why they took that approach but I would not be surprised if we can achieve the same without needing to return a `PCollectionTuple` in the Write.

Reply:
@iemejia , I think we are not in synced. Of course, I understand your point. But, like I explained above, using PCollection<Void> or PDone as return type, will not give us any way to return any result to user. 

> @iemejia , I have changed it to use PCollection<Void>, but couldn't find a way to return multiple outputs. Something similar to this JDBC one, https://github.com/apache/beam/blob/master/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java#L842, it also return just one output if using PCollection<Void>. Any idea of how this would be done?